File: packages/workflow-framework/src/governance/work-id.ts
/**
 * Work ID governance utilities.
 *
 * Work IDs are the single most important traceability primitive in the Golden
 * Path. Every commit, branch, PR title, and audit event must carry a Work ID
 * so that deployments can be traced back to planned work.
 *
 * Conventions enforced here:
 *   - Default pattern: ^[A-Z]+-[0-9]+$  (e.g. FIN-123, PLAT-456)
 *   - Branch pattern:  ^(feature|fix|chore|hotfix)/[A-Z]+-[0-9]+-[a-z0-9-]+$
 *   - Commit pattern:  ^\[[A-Z]+-[0-9]+\] .+
 *   - PR title:        ^\[[A-Z]+-[0-9]+\] .+
 *
 * All patterns are configurable via devex.yaml (workTracking section) so that
 * teams using Linear, Shortcut, or custom trackers can adapt.
 */

export interface WorkIdConfig {
  /** Regex string that a bare Work ID must satisfy. */
  workIdPattern: string;
  /** Regex string that a branch name must satisfy. */
  branchPattern: string;
  /** Regex string that a commit message must satisfy. */
  commitPattern: string;
  /** Regex string that a PR title must satisfy. */
  prTitlePattern: string;
}

/** Defaults that mirror the devex.yaml schema defaults. */
export const DEFAULT_WORK_ID_CONFIG: WorkIdConfig = {
  workIdPattern: "^[A-Z]+-[0-9]+$",
  branchPattern: "^(feature|fix|chore|hotfix)/[A-Z]+-[0-9]+-[a-z0-9-]+$",
  commitPattern: "^\\[[A-Z]+-[0-9]+\\] .+",
  prTitlePattern: "^\\[[A-Z]+-[0-9]+\\] .+",
};

// ---------------------------------------------------------------------------
// Core validators
// ---------------------------------------------------------------------------

export interface WorkIdValidationResult {
  valid: boolean;
  workId: string | null;
  message: string;
}

/**
 * Validates a bare Work ID string against the configured pattern.
 *
 * @example
 * validateWorkId("FIN-123")  → { valid: true, workId: "FIN-123" }
 * validateWorkId("fin123")   → { valid: false, workId: null, message: "..." }
 */
export function validateWorkId(
  value: string,
  config: WorkIdConfig = DEFAULT_WORK_ID_CONFIG
): WorkIdValidationResult {
  const pattern = new RegExp(config.workIdPattern);

  if (pattern.test(value.trim())) {
    return { valid: true, workId: value.trim(), message: "Work ID is valid." };
  }

  return {
    valid: false,
    workId: null,
    message:
      `"${value}" does not match the required Work ID pattern: ${config.workIdPattern}. ` +
      `Expected format: PROJECT-NUMBER (e.g. FIN-123).`,
  };
}

/**
 * Validates a branch name against the configured branch pattern.
 * Also extracts the Work ID embedded in the branch name when valid.
 *
 * @example
 * validateBranchName("feature/FIN-123-add-validation")
 *   → { valid: true, workId: "FIN-123" }
 *
 * validateBranchName("feature/add-validation")
 *   → { valid: false, workId: null, message: "..." }
 */
export function validateBranchName(
  branchName: string,
  config: WorkIdConfig = DEFAULT_WORK_ID_CONFIG
): WorkIdValidationResult {
  const pattern = new RegExp(config.branchPattern);

  if (!pattern.test(branchName)) {
    return {
      valid: false,
      workId: null,
      message:
        `Branch name "${branchName}" does not match the required Work ID branch pattern:\n` +
        `  ${config.branchPattern}\n` +
        `Expected format: <type>/<WORK-ID>-<description> ` +
        `(e.g. feature/FIN-123-add-transaction-validation).`,
    };
  }

  const workId = extractWorkIdFromBranch(branchName, config);
  return {
    valid: true,
    workId,
    message: workId
      ? `Branch contains Work ID ${workId}.`
      : "Branch is valid but no Work ID could be extracted.",
  };
}

/**
 * Validates a commit message against the configured commit pattern.
 *
 * @example
 * validateCommitMessage("[FIN-123] Add transaction validation")
 *   → { valid: true, workId: "FIN-123" }
 */
export function validateCommitMessage(
  message: string,
  config: WorkIdConfig = DEFAULT_WORK_ID_CONFIG
): WorkIdValidationResult {
  const firstLine = message.split("\n")[0]?.trim() ?? "";
  const pattern = new RegExp(config.commitPattern);

  if (!pattern.test(firstLine)) {
    return {
      valid: false,
      workId: null,
      message:
        `Commit message "${firstLine}" does not match the required Work ID commit pattern:\n` +
        `  ${config.commitPattern}\n` +
        `Expected format: [WORK-ID] Description (e.g. [FIN-123] Add transaction validation).`,
    };
  }

  const workId = extractWorkIdFromText(firstLine, config);
  return {
    valid: true,
    workId,
    message: workId
      ? `Commit message contains Work ID ${workId}.`
      : "Commit message is valid.",
  };
}

/**
 * Validates a PR title against the configured PR title pattern.
 *
 * @example
 * validatePrTitle("[FIN-123] Add transaction validation")
 *   → { valid: true, workId: "FIN-123" }
 */
export function validatePrTitle(
  title: string,
  config: WorkIdConfig = DEFAULT_WORK_ID_CONFIG
): WorkIdValidationResult {
  const pattern = new RegExp(config.prTitlePattern);

  if (!pattern.test(title.trim())) {
    return {
      valid: false,
      workId: null,
      message:
        `PR title "${title}" does not match the required pattern:\n` +
        `  ${config.prTitlePattern}\n` +
        `Expected format: [WORK-ID] Description (e.g. [FIN-123] Add transaction validation).`,
    };
  }

  const workId = extractWorkIdFromText(title, config);
  return {
    valid: true,
    workId,
    message: workId
      ? `PR title contains Work ID ${workId}.`
      : "PR title is valid.",
  };
}

// ---------------------------------------------------------------------------
// Batch validation
// ---------------------------------------------------------------------------

export interface CommitValidationSummary {
  /** Total commits evaluated. */
  total: number;
  /** Commits that passed the pattern check. */
  passed: number;
  /** Commit messages (first lines) that failed the pattern check. */
  violations: string[];
}

/**
 * Validates a list of commit messages and returns a summary.
 * Use this in the governance job to check all commits in a PR.
 */
export function validateCommitMessages(
  messages: string[],
  config: WorkIdConfig = DEFAULT_WORK_ID_CONFIG
): CommitValidationSummary {
  const violations: string[] = [];
  let passed = 0;

  for (const message of messages) {
    const result = validateCommitMessage(message, config);
    if (result.valid) {
      passed++;
    } else {
      violations.push(message.split("\n")[0]?.trim() ?? message);
    }
  }

  return { total: messages.length, passed, violations };
}

// ---------------------------------------------------------------------------
// Extraction helpers
// ---------------------------------------------------------------------------

/**
 * Extracts a Work ID from a branch name using the workIdPattern.
 * Returns null if no Work ID is found.
 *
 * @example
 * extractWorkIdFromBranch("feature/FIN-123-add-validation") → "FIN-123"
 */
export function extractWorkIdFromBranch(
  branchName: string,
  config: WorkIdConfig = DEFAULT_WORK_ID_CONFIG
): string | null {
  return extractWorkIdFromText(branchName, config);
}

/**
 * Extracts a Work ID from arbitrary text using the workIdPattern.
 * Uses the first match found. Returns null if no match.
 */
export function extractWorkIdFromText(
  text: string,
  config: WorkIdConfig = DEFAULT_WORK_ID_CONFIG
): string | null {
  // Strip regex anchors for use as a search pattern within larger strings
  const innerPattern = config.workIdPattern
    .replace(/^\^/, "")
    .replace(/\$$/, "");

  const searchRegex = new RegExp(innerPattern, "g");
  const match = searchRegex.exec(text);
  return match?.[0] ?? null;
}

// ---------------------------------------------------------------------------
// Branch name builder (used by CLI `devex branch`)
// ---------------------------------------------------------------------------

export type BranchType = "feature" | "fix" | "chore" | "hotfix";

/**
 * Builds a compliant branch name from a type, Work ID, and description.
 *
 * @example
 * buildBranchName("feature", "FIN-123", "add transaction validation")
 *   → "feature/FIN-123-add-transaction-validation"
 */
export function buildBranchName(
  type: BranchType,
  workId: string,
  description: string
): string {
  const slug = description
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9\s-]/g, "")
    .replace(/\s+/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");

  return `${type}/${workId}-${slug}`;
}

/**
 * Builds a compliant commit message first line.
 *
 * @example
 * buildCommitMessage("FIN-123", "add transaction validation")
 *   → "[FIN-123] add transaction validation"
 */
export function buildCommitMessage(workId: string, description: string): string {
  return `[${workId}] ${description.trim()}`;
}

/**
 * Builds a compliant PR title.
 *
 * @example
 * buildPrTitle("FIN-123", "Add transaction validation")
 *   → "[FIN-123] Add transaction validation"
 */
export function buildPrTitle(workId: string, title: string): string {
  return `[${workId}] ${title.trim()}`;
}

File: packages/cli/src/devex_cli/governance/work_id.py
"""Work ID governance utilities — Python mirror of governance/work-id.ts.

Enforces the same branch-name, commit-message, and Work ID rules locally
that the workflow-framework CI pipeline enforces remotely.
"""

from __future__ import annotations

import re
import subprocess
from dataclasses import dataclass
from typing import Optional

from devex_cli.config.schema import WorkTrackingConfig

# ---------------------------------------------------------------------------
# Result type
# ---------------------------------------------------------------------------


@dataclass
class WorkIdResult:
    valid: bool
    work_id: Optional[str]
    message: str


# ---------------------------------------------------------------------------
# Core validators
# ---------------------------------------------------------------------------


def validate_work_id(value: str, config: WorkTrackingConfig = WorkTrackingConfig()) -> WorkIdResult:
    """Validate a bare Work ID string against the configured pattern."""
    if re.match(config.workIdPattern, value.strip()):
        return WorkIdResult(valid=True, work_id=value.strip(), message="Work ID is valid.")
    return WorkIdResult(
        valid=False,
        work_id=None,
        message=(
            f'"{value}" does not match the required Work ID pattern: {config.workIdPattern}. '
            "Expected format: PROJECT-NUMBER (e.g. FIN-123)."
        ),
    )


def validate_branch_name(
    branch: str, config: WorkTrackingConfig = WorkTrackingConfig()
) -> WorkIdResult:
    """Validate a branch name and extract the embedded Work ID."""
    if not re.match(config.branchPattern, branch):
        return WorkIdResult(
            valid=False,
            work_id=None,
            message=(
                f'Branch name "{branch}" does not match the required pattern:\n'
                f"  {config.branchPattern}\n"
                "Expected format: <type>/<WORK-ID>-<description> "
                "(e.g. feature/FIN-123-add-transaction-validation)."
            ),
        )
    work_id = _extract_work_id_from_branch(branch, config)
    msg = f"Branch contains Work ID {work_id}." if work_id else "Branch is valid."
    return WorkIdResult(valid=True, work_id=work_id, message=msg)


def validate_commit_message(
    message: str, config: WorkTrackingConfig = WorkTrackingConfig()
) -> WorkIdResult:
    """Validate the first line of a commit message."""
    first_line = message.split("\n")[0].strip()
    if not re.match(config.commitPattern, first_line):
        return WorkIdResult(
            valid=False,
            work_id=None,
            message=(
                f'Commit message "{first_line}" does not match the required pattern:\n'
                f"  {config.commitPattern}\n"
                "Expected format: [WORK-ID] Description (e.g. [FIN-123] Add validation)."
            ),
        )
    work_id = _extract_work_id_from_text(first_line, config)
    msg = f"Commit message contains Work ID {work_id}." if work_id else "Commit message is valid."
    return WorkIdResult(valid=True, work_id=work_id, message=msg)


# ---------------------------------------------------------------------------
# Git helpers
# ---------------------------------------------------------------------------


def current_branch() -> Optional[str]:
    """Return the current Git branch name, or None when not in a Git repo."""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True,
            text=True,
            check=True,
        )
        branch = result.stdout.strip()
        return branch if branch and branch != "HEAD" else None
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None


def recent_commits(n: int = 5) -> list[str]:
    """Return the first lines of the *n* most recent commit messages."""
    try:
        result = subprocess.run(
            ["git", "log", f"--max-count={n}", "--format=%s"],
            capture_output=True,
            text=True,
            check=True,
        )
        return [line for line in result.stdout.splitlines() if line.strip()]
    except (subprocess.CalledProcessError, FileNotFoundError):
        return []


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _strip_anchors(pattern: str) -> str:
    """Remove leading ^ and trailing $ anchors so the pattern can match within a longer string."""
    return pattern.lstrip("^").rstrip("$")


def _extract_work_id_from_branch(branch: str, config: WorkTrackingConfig) -> Optional[str]:
    """Pull the Work ID segment out of a branch name like feature/FIN-123-foo."""
    match = re.search(_strip_anchors(config.workIdPattern), branch)
    return match.group(0) if match else None


def _extract_work_id_from_text(text: str, config: WorkTrackingConfig) -> Optional[str]:
    """Pull the Work ID out of a commit message or PR title."""
    match = re.search(_strip_anchors(config.workIdPattern), text)
    return match.group(0) if match else None
